x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/index/store/cache/ProgressListenableActionFutureTests.java

```diff
@@ -0,0 +1,172 @@
+        final ProgressListenableActionFuture future = randomFuture();
+
+        final AtomicArray<Exception> listenersResponses = new AtomicArray<>(between(0, 50));
+        for (int i = 0; i < listenersResponses.length(); i++) {
+            final int listenerIndex = i;
+            future.addListener(
+                ActionListener.wrap(
+                    o -> listenersResponses.setOnce(listenerIndex, null),
+                    e -> listenersResponses.setOnce(listenerIndex, e)
+                ),
+                randomLongBetween(future.start, future.end)
+            );
+        }
+        assertTrue(listenersResponses.asList().stream().allMatch(Objects::isNull));
+
+        final Exception exception = new ElasticsearchException("simulated");
+        future.onFailure(exception);
+
+        for (int i = 0; i < listenersResponses.length(); i++) {
+            assertThat(listenersResponses.get(i), sameInstance(exception));
+        }
+
+        IllegalStateException ise = expectThrows(IllegalStateException.class, () -> future.onFailure(exception));
+        assertThat(ise.getMessage(), containsString("Future is already completed"));
+    }
+
+    public void testProgressUpdatesCallsListeners() throws Exception {
+        final ProgressListenableActionFuture future = randomFuture();
+
+        final Thread[] threads = new Thread[between(1, 5)];
+        final CountDownLatch startLatch = new CountDownLatch(1);
+
+        for (int i = 0; i < threads.length; i++) {
+            threads[i] = new Thread(() -> {
+                try {
+                    startLatch.await();
+                    while (future.isDone() == false) {
+                        final long expectedProgress = randomLongBetween(future.start, future.end);
+                        final PlainActionFuture<Long> listener = new PlainActionFuture<>();
+                        future.addListener(ActionListener.wrap(listener::onResponse, listener::onFailure), expectedProgress);
+                        assertThat(listener.get(), greaterThanOrEqualTo(expectedProgress));
+                    }
+                } catch (Throwable t) {
+                    logger.error("Failed to wait for progress to be reached", t);
+                    if (future.isDone() == false) {
+                        future.onFailure(
+                            new Exception("Failed to update progress [" + t.getClass().getName() + ':' + t.getMessage() + "]")
+                        );
+                    }
+                    throw new AssertionError(t);
+                }
+            });
+        }
+
+        for (Thread thread : threads) {
+            thread.start();
+        }
+
+        final Thread progressUpdaterThread = new Thread(() -> {
+            try {
+                startLatch.await();
+                long progress = future.start;
+                while (progress < future.end) {
+                    progress += randomLongBetween(1L, Math.max(1L, future.end - progress));
```

Comment:
Perhaps simpler:

```suggestion
                    progress += randomLongBetween(progress + 1L, future.end));
```

Reply:
I'll go with `progress = randomLongBetween(1L, Math.max(1L, future.end - progress);` - thanks for the suggestion